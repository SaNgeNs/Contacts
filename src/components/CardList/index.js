import withErrorHandler from 'Components/withErrorHandler';
import loadable from '@loadable/component';

const CardList = loadable(() => import(/* webpackChunkName: "card_list_component" */'./CardList'));

export default withErrorHandler(CardList);
