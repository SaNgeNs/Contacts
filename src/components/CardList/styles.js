export const styles = {
  wrap: {
    display: 'flex',
    flexWrap: 'wrap',
  },
  card: {
    margin: '10px',
    width: '340px',
    flexShrink: 0,
  },
  modalContent: {
    padding: '20px',
    background: '#fff',
    boxShadow: '0px 3px 5px -1px rgba(0,0,0,0.2), 0px 5px 8px 0px rgba(0,0,0,0.14), 0px 1px 14px 0px rgba(0,0,0,0.12)',
    wordBreak: 'break-all',
    margin: '0 20px',
    outline: 'none',
  },
  modal: {
    display: 'flex',
    justifyContent: 'center',
    alignItems: 'center',
  },
};

export default styles;
