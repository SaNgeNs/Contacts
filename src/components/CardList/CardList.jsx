import React, { memo, useState } from 'react';
import withStyles from '@material-ui/core/styles/withStyles';
import Card from '@material-ui/core/Card';
import CardContent from '@material-ui/core/CardContent';
import CardActions from '@material-ui/core/CardActions';
import Button from '@material-ui/core/Button';
import IconButton from '@material-ui/core/IconButton';
import Modal from '@material-ui/core/Modal';
import Backdrop from '@material-ui/core/Backdrop';
import Fade from '@material-ui/core/Fade';
import FavoriteIcon from '@material-ui/icons/Favorite';
import FavoriteBorder from '@material-ui/icons/FavoriteBorder';
import { useDispatch } from 'react-redux';

import toggleFavourite from 'Actions/toggleFavourite';
import EditForm from 'Components/Forms/EditForm';
import styles from './styles';

export const CardList = ({
  classes,
  cards = [],
}) => {
  const dispatch = useDispatch();
  const [ openModal, setOpenModal ] = useState(false);
  const [ editCard, setEditCard ] = useState({});
  const onClose = () => {
    setOpenModal(false);
    setEditCard({});
  };

  return (
    <div className={classes.wrap}>
      {cards.map(card => (
        <Card className={classes.card} key={card.id}>
          <CardContent>Name: {card.name}</CardContent>
          <CardContent>Phone: {card.phone}</CardContent>
          <CardContent>Email: {card.email}</CardContent>

          <CardActions>
            <IconButton
              aria-label="add to favorites"
              onClick={() => {
                dispatch(toggleFavourite({
                  id: card.id,
                  isFavourite: !card.isFavourite,
                }))
              }}
            >
              {card.isFavourite ? (
                <FavoriteIcon />
              ) : (
                <FavoriteBorder />
              )}
            </IconButton>

            <Button
              aria-label="edit card"
              variant="contained"
              onClick={() => {
                setOpenModal(!openModal);
                setEditCard(card);
              }}
            >
              Edit
            </Button>
          </CardActions>
        </Card>
      ))}

      <Modal
        className={classes.modal}
        open={openModal}
        onClose={onClose}
        closeAfterTransition
        BackdropComponent={Backdrop}
        BackdropProps={{
          timeout: 500,
        }}
      >
        <Fade in={openModal}>
          <div className={classes.modalContent}>
            <EditForm
              cardInfo={editCard}
              onClickAfterSubmit={onClose}
              initialValues={{
                name: editCard.name,
                phone: editCard.phone,
                email: editCard.email,
              }}
            />
          </div>
        </Fade>
      </Modal>
    </div>
  );
};

export default memo(withStyles(styles)(CardList));
