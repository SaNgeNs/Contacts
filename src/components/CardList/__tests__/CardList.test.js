import React from 'react';
import { fireEvent } from '@testing-library/react';
import customRender from 'Mock/customRender';
import MockCards from 'Mock/cards';

import CardList from '../CardList';

describe('<CardList />', () => {
  it('should render card list: ', () => {
    const { container, unmount } = customRender(
      <CardList
        cards={MockCards}
      />
    );

    expect(container).toMatchSnapshot();
    unmount();
  });

  it('add favourite card: ', () => {
    const { container, unmount } = customRender(
      <CardList
        cards={[MockCards[0]]}
      />
    );

    const button = container.querySelector('[aria-label="add to favorites"]');

    fireEvent.click(button);
    expect(container).toMatchSnapshot();
    unmount();
  });

  it('should render edit form: ', async () => {
    const { container, unmount, queryByRole, getByRole } = customRender(
      <CardList
        cards={[MockCards[0]]}
      />
    );

    const button = container.querySelectorAll('[aria-label="edit card"]')[0];

    fireEvent.click(button);
    await new Promise(resolve => setTimeout(resolve, 600));
    expect(container).toMatchSnapshot();

    const presentation = getByRole('presentation');
    const modalBg = presentation.querySelectorAll('div')[0];
    expect(presentation).toBeInTheDocument();
    expect(container).toMatchSnapshot();

    fireEvent.click(modalBg);
    await new Promise(resolve => setTimeout(resolve, 600));
    expect(queryByRole('presentation')).toBe(null);
    expect(container).toMatchSnapshot();

    unmount();
  });

  it('should render empty cards: ', () => {
    const { container, unmount } = customRender(
      <CardList />
    );

    expect(container.childElementCount).toBe(1);
    expect(container).toMatchSnapshot();
    unmount();
  });
});
