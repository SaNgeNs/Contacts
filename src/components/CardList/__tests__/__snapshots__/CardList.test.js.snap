// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CardList /> add favourite card:  1`] = `
<div>
  <div
    class="CardList-wrap-5"
  >
    <div
      class="MuiPaper-root MuiCard-root CardList-card-6 MuiPaper-elevation1 MuiPaper-rounded"
    >
      <div
        class="MuiCardContent-root"
      >
        Name: 
        test1
      </div>
      <div
        class="MuiCardContent-root"
      >
        Phone: 
        1234567890
      </div>
      <div
        class="MuiCardContent-root"
      >
        Email: 
        test@mail.ru
      </div>
      <div
        class="MuiCardActions-root MuiCardActions-spacing"
      >
        <button
          aria-label="add to favorites"
          class="MuiButtonBase-root MuiIconButton-root"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiIconButton-label"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M16.5 3c-1.74 0-3.41.81-4.5 2.09C10.91 3.81 9.24 3 7.5 3 4.42 3 2 5.42 2 8.5c0 3.78 3.4 6.86 8.55 11.54L12 21.35l1.45-1.32C18.6 15.36 22 12.28 22 8.5 22 5.42 19.58 3 16.5 3zm-4.4 15.55l-.1.1-.1-.1C7.14 14.24 4 11.39 4 8.5 4 6.5 5.5 5 7.5 5c1.54 0 3.04.99 3.57 2.36h1.87C13.46 5.99 14.96 5 16.5 5c2 0 3.5 1.5 3.5 3.5 0 2.89-3.14 5.74-7.9 10.05z"
              />
            </svg>
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
        <button
          aria-label="edit card"
          class="MuiButtonBase-root MuiButton-root MuiButton-contained"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiButton-label"
          >
            Edit
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`<CardList /> should render card list:  1`] = `
<div>
  <div
    class="CardList-wrap-1"
  >
    <div
      class="MuiPaper-root MuiCard-root CardList-card-2 MuiPaper-elevation1 MuiPaper-rounded"
    >
      <div
        class="MuiCardContent-root"
      >
        Name: 
        test1
      </div>
      <div
        class="MuiCardContent-root"
      >
        Phone: 
        1234567890
      </div>
      <div
        class="MuiCardContent-root"
      >
        Email: 
        test@mail.ru
      </div>
      <div
        class="MuiCardActions-root MuiCardActions-spacing"
      >
        <button
          aria-label="add to favorites"
          class="MuiButtonBase-root MuiIconButton-root"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiIconButton-label"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M16.5 3c-1.74 0-3.41.81-4.5 2.09C10.91 3.81 9.24 3 7.5 3 4.42 3 2 5.42 2 8.5c0 3.78 3.4 6.86 8.55 11.54L12 21.35l1.45-1.32C18.6 15.36 22 12.28 22 8.5 22 5.42 19.58 3 16.5 3zm-4.4 15.55l-.1.1-.1-.1C7.14 14.24 4 11.39 4 8.5 4 6.5 5.5 5 7.5 5c1.54 0 3.04.99 3.57 2.36h1.87C13.46 5.99 14.96 5 16.5 5c2 0 3.5 1.5 3.5 3.5 0 2.89-3.14 5.74-7.9 10.05z"
              />
            </svg>
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
        <button
          aria-label="edit card"
          class="MuiButtonBase-root MuiButton-root MuiButton-contained"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiButton-label"
          >
            Edit
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
    <div
      class="MuiPaper-root MuiCard-root CardList-card-2 MuiPaper-elevation1 MuiPaper-rounded"
    >
      <div
        class="MuiCardContent-root"
      >
        Name: 
        qwerty
      </div>
      <div
        class="MuiCardContent-root"
      >
        Phone: 
        12345678909
      </div>
      <div
        class="MuiCardContent-root"
      >
        Email: 
        test2@mail.ru
      </div>
      <div
        class="MuiCardActions-root MuiCardActions-spacing"
      >
        <button
          aria-label="add to favorites"
          class="MuiButtonBase-root MuiIconButton-root"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiIconButton-label"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M12 21.35l-1.45-1.32C5.4 15.36 2 12.28 2 8.5 2 5.42 4.42 3 7.5 3c1.74 0 3.41.81 4.5 2.09C13.09 3.81 14.76 3 16.5 3 19.58 3 22 5.42 22 8.5c0 3.78-3.4 6.86-8.55 11.54L12 21.35z"
              />
            </svg>
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
        <button
          aria-label="edit card"
          class="MuiButtonBase-root MuiButton-root MuiButton-contained"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiButton-label"
          >
            Edit
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`<CardList /> should render edit form:  1`] = `
<div
  aria-hidden="true"
>
  <div
    class="CardList-wrap-9"
  >
    <div
      class="MuiPaper-root MuiCard-root CardList-card-10 MuiPaper-elevation1 MuiPaper-rounded"
    >
      <div
        class="MuiCardContent-root"
      >
        Name: 
        test1
      </div>
      <div
        class="MuiCardContent-root"
      >
        Phone: 
        1234567890
      </div>
      <div
        class="MuiCardContent-root"
      >
        Email: 
        test@mail.ru
      </div>
      <div
        class="MuiCardActions-root MuiCardActions-spacing"
      >
        <button
          aria-label="add to favorites"
          class="MuiButtonBase-root MuiIconButton-root"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiIconButton-label"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M16.5 3c-1.74 0-3.41.81-4.5 2.09C10.91 3.81 9.24 3 7.5 3 4.42 3 2 5.42 2 8.5c0 3.78 3.4 6.86 8.55 11.54L12 21.35l1.45-1.32C18.6 15.36 22 12.28 22 8.5 22 5.42 19.58 3 16.5 3zm-4.4 15.55l-.1.1-.1-.1C7.14 14.24 4 11.39 4 8.5 4 6.5 5.5 5 7.5 5c1.54 0 3.04.99 3.57 2.36h1.87C13.46 5.99 14.96 5 16.5 5c2 0 3.5 1.5 3.5 3.5 0 2.89-3.14 5.74-7.9 10.05z"
              />
            </svg>
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
        <button
          aria-label="edit card"
          class="MuiButtonBase-root MuiButton-root MuiButton-contained"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiButton-label"
          >
            Edit
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`<CardList /> should render edit form:  2`] = `
<div
  aria-hidden="true"
>
  <div
    class="CardList-wrap-9"
  >
    <div
      class="MuiPaper-root MuiCard-root CardList-card-10 MuiPaper-elevation1 MuiPaper-rounded"
    >
      <div
        class="MuiCardContent-root"
      >
        Name: 
        test1
      </div>
      <div
        class="MuiCardContent-root"
      >
        Phone: 
        1234567890
      </div>
      <div
        class="MuiCardContent-root"
      >
        Email: 
        test@mail.ru
      </div>
      <div
        class="MuiCardActions-root MuiCardActions-spacing"
      >
        <button
          aria-label="add to favorites"
          class="MuiButtonBase-root MuiIconButton-root"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiIconButton-label"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M16.5 3c-1.74 0-3.41.81-4.5 2.09C10.91 3.81 9.24 3 7.5 3 4.42 3 2 5.42 2 8.5c0 3.78 3.4 6.86 8.55 11.54L12 21.35l1.45-1.32C18.6 15.36 22 12.28 22 8.5 22 5.42 19.58 3 16.5 3zm-4.4 15.55l-.1.1-.1-.1C7.14 14.24 4 11.39 4 8.5 4 6.5 5.5 5 7.5 5c1.54 0 3.04.99 3.57 2.36h1.87C13.46 5.99 14.96 5 16.5 5c2 0 3.5 1.5 3.5 3.5 0 2.89-3.14 5.74-7.9 10.05z"
              />
            </svg>
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
        <button
          aria-label="edit card"
          class="MuiButtonBase-root MuiButton-root MuiButton-contained"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiButton-label"
          >
            Edit
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`<CardList /> should render edit form:  3`] = `
<div>
  <div
    class="CardList-wrap-9"
  >
    <div
      class="MuiPaper-root MuiCard-root CardList-card-10 MuiPaper-elevation1 MuiPaper-rounded"
    >
      <div
        class="MuiCardContent-root"
      >
        Name: 
        test1
      </div>
      <div
        class="MuiCardContent-root"
      >
        Phone: 
        1234567890
      </div>
      <div
        class="MuiCardContent-root"
      >
        Email: 
        test@mail.ru
      </div>
      <div
        class="MuiCardActions-root MuiCardActions-spacing"
      >
        <button
          aria-label="add to favorites"
          class="MuiButtonBase-root MuiIconButton-root"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiIconButton-label"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M16.5 3c-1.74 0-3.41.81-4.5 2.09C10.91 3.81 9.24 3 7.5 3 4.42 3 2 5.42 2 8.5c0 3.78 3.4 6.86 8.55 11.54L12 21.35l1.45-1.32C18.6 15.36 22 12.28 22 8.5 22 5.42 19.58 3 16.5 3zm-4.4 15.55l-.1.1-.1-.1C7.14 14.24 4 11.39 4 8.5 4 6.5 5.5 5 7.5 5c1.54 0 3.04.99 3.57 2.36h1.87C13.46 5.99 14.96 5 16.5 5c2 0 3.5 1.5 3.5 3.5 0 2.89-3.14 5.74-7.9 10.05z"
              />
            </svg>
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
        <button
          aria-label="edit card"
          class="MuiButtonBase-root MuiButton-root MuiButton-contained"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiButton-label"
          >
            Edit
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`<CardList /> should render empty cards:  1`] = `
<div>
  <div
    class="CardList-wrap-20"
  />
</div>
`;
